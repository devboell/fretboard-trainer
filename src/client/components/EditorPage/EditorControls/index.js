import React from 'react'
import pt from 'prop-types'

import { quizTypes } from 'components/EditorPage/ListEditor/EditorContainer/constants'
import NewButton from './NewButton'
import PreviewButton from './PreviewButton'
import Wrapper from './Wrapper'


const EditorControls = ({
  onSelectNewItem,
  onOpenPreview,
  isNew,
  hasSelection,
}) =>
  <Wrapper>
    <NewButton
      clickAction={onSelectNewItem}
      items={quizTypes}
      disabled={isNew}
    >
      New
    </NewButton>
    <PreviewButton
      onClick={() => onOpenPreview()}
      disabled={!hasSelection}
    >
      Preview
    </PreviewButton>
  </Wrapper>

EditorControls.propTypes = {
  onSelectNewItem: pt.func.isRequired,
  onOpenPreview: pt.func.isRequired,
  isNew: pt.bool.isRequired,
  hasSelection: pt.bool.isRequired,
}


export default EditorControls
