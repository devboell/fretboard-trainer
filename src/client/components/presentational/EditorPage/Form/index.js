import React from 'react'
import pt from 'prop-types'
import { compose, partialRight, propEq, prop, cond, T } from 'ramda'
import { quizShape } from 'propShapes/quiz'

const parseInteger = partialRight(parseInt, [10])

export const FormContext = React.createContext()

class Form extends React.Component {
  constructor(props) {
    super(props)

    this.handleInputChange = this.handleInputChange.bind(this)
    this.handleCheckBoxGroupChange = this.handleCheckBoxGroupChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleDelete = this.handleDelete.bind(this)
  }

  handleInputChange(event) {
    const { onUpdateBuffer } = this.props
    const { target, target: { name } } = event
    const typeEquals = propEq('type')
    const value = cond([
      [typeEquals('checkbox'), prop('checked')],
      [typeEquals('number'), compose(parseInteger, prop('value'))],
      [T, prop('value')],
    ])(target)

    onUpdateBuffer(name, value, 'SET')
  }

  handleCheckBoxGroupChange(event) {
    const { onUpdateBuffer } = this.props
    const { target } = event
    const { name, value, checked } = target
    const updateType = checked ? 'ADD' : 'REMOVE'

    onUpdateBuffer(name, value, updateType)
  }

  handleSubmit(event) {
    event.preventDefault()
    const {
      isNew,
      buffer,
      onCreate,
      onUpdate,
    } = this.props

    isNew
      ? onCreate(buffer)
      : onUpdate(buffer)
  }

  handleDelete(event) {
    const { onDelete, buffer } = this.props
    event.preventDefault()

    onDelete({ id: buffer.id })
  }

  render() {
    const {
      buffer,
      isPristine,
      isNew,
      className,
    } = this.props
    const context = {
      buffer,
      isPristine,
      isNew,
      handleInputChange: this.handleInputChange,
      handleCheckBoxGroupChange: this.handleCheckBoxGroupChange,
      handleDelete: this.handleDelete,
    }
    return (
      <form
        className={className}
        onSubmit={this.handleSubmit}
      >
        <FormContext.Provider value={context}>
          {this.props.children}
        </FormContext.Provider>
      </form>
    )
  }
}


Form.propTypes = {
  isNew: pt.bool.isRequired,
  buffer: quizShape.isRequired,
  onUpdateBuffer: pt.func.isRequired,
  isPristine: pt.bool.isRequired,
  onCreate: pt.func.isRequired,
  onUpdate: pt.func.isRequired,
  onDelete: pt.func.isRequired,
  className: pt.string.isRequired,
  children: pt.node.isRequired,
}


export default Form
