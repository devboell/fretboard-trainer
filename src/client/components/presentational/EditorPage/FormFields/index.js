import React from 'react'
import pt from 'prop-types'
import { filter, contains } from 'ramda'

import { pcPanelModeIds } from 'components/containers/QuizEditor/constants'

import Wrapper from './Wrapper'

import NameField from './NameField'
import TypeField from './TypeField'
import DescriptionField from './DescriptionField'
import PanelModeField from './PanelModeField'
import AllAnswersField from './AllAnswersField'
import AllowIncorrectField from './AllowIncorrectField'
import TimerField from './TimerField'
import ShowNotesField from './ShowNotesField'
import HighlightOctavesField from './HighlightOctavesField'


const panelModesForType = (type, panelModes) => (
  type === 'pc'
    ? filter(pm => contains(pm.id, pcPanelModeIds), panelModes)
    : panelModes)

const fieldsByType = {
  pc: [],
  pitch: [
    <HighlightOctavesField key="highlightOctavesField" />,
  ],
  interval: [],
}

const fields = (panelModes, type) => [
  <NameField key="nameField" />,
  <TypeField key="typeField" />,
  <DescriptionField key="descriptionField" />,
  <PanelModeField
    key="panelField"
    panelModes={panelModesForType(type, panelModes)}
  />,
  <AllAnswersField key="allAnswersField" />,
  <AllowIncorrectField key="allowIncorrectField" />,
  <TimerField key="timerField" />,
  <ShowNotesField key="showNotesField" />,
  ...fieldsByType[type],
]

const FormFields = ({ panelModes, quizType }) =>
  <Wrapper>
    {fields(panelModes, quizType)}
  </Wrapper>

FormFields.propTypes = {
  panelModes: pt.arrayOf(pt.shape({})).isRequired,
  quizType: pt.string.isRequired,
}

export default FormFields
