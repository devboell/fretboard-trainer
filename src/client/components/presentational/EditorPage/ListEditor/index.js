import React from 'react'
import pt from 'prop-types'
import { connect } from 'react-redux'
import { stopQuiz } from 'components/containers/Runner'

import Preview from 'components/presentational/EditorPage/Preview'
import QuizList from 'components/containers/QuizList'

import { togglePreview } from './EditorContainer/reducer'
import Wrapper from './Wrapper'
import ListEditorWrapper from './ListEditorWrapper'
import EditorContainer from './EditorContainer'


const ListEditor = ({ showPreview, onClosePreview }) =>
  <Wrapper>
    {showPreview
      ? <Preview {...{ onClosePreview }} />
      :
      <ListEditorWrapper>
        <QuizList />
        <EditorContainer />
      </ListEditorWrapper>
    }
  </Wrapper>

ListEditor.propTypes = {
  showPreview: pt.bool.isRequired,
  onClosePreview: pt.func.isRequired,
}

const mapStateToProps = state => ({
  showPreview: state.editor.showPreview,
})

const closePreview = (dispatch) => {
  dispatch(stopQuiz())
  dispatch(togglePreview())
}

const mapDispatchToProps = dispatch => ({
  onClosePreview: () => closePreview(dispatch),
})

export default connect(mapStateToProps, mapDispatchToProps)(ListEditor)
