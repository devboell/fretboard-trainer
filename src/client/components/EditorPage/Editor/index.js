import React from 'react'
import pt from 'prop-types'
import { equals } from 'ramda'
import { quizShape } from 'propShapes/quiz'
import { modes } from 'components/EditorPage/ListEditor/EditorContainer/reducer'

import FormFields from 'components/EditorPage/FormFields'
import FormControls from 'components/EditorPage/FormControls'
import EditorControls from 'components/EditorPage/EditorControls'

import UnselectedMessage from './UnselectedMessage'

import Wrapper from './Wrapper'
import StyledForm from './StyledForm'

const Editor = ({
  mode,
  original,
  buffer,
  panelModes,
  onSelectNewItem,
  onOpenPreview,
  ...rest
}) => {
  const isNew = mode === modes.NEW
  const isPristine = equals(original, buffer)
  const hasSelection = mode !== modes.UNSELECTED

  return (
    <Wrapper>
      <EditorControls
        {...{
          onSelectNewItem, onOpenPreview, isNew, hasSelection,
        }}
      />
      {mode === modes.UNSELECTED
        ? <UnselectedMessage />
        :
        <StyledForm
          {...{
            buffer,
            isNew,
            isPristine,
            ...rest,
          }}
        >
          <FormFields {...{ panelModes }} />
          <FormControls />
        </StyledForm>
      }
    </Wrapper>
  )
}


Editor.propTypes = {
  mode: pt.string.isRequired,
  original: quizShape,
  buffer: quizShape,
  panelModes: pt.arrayOf(pt.shape({})).isRequired,
  onUpdateBuffer: pt.func.isRequired,
  onSelectNewItem: pt.func.isRequired,
  onOpenPreview: pt.func.isRequired,
  onCreate: pt.func.isRequired,
  onUpdate: pt.func.isRequired,
  onDelete: pt.func.isRequired,
}

Editor.defaultProps = {
  original: undefined,
  buffer: undefined,
}

export default Editor
