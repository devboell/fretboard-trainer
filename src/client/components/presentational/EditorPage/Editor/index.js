import React from 'react'
import pt from 'prop-types'
import { equals, filter, contains } from 'ramda'
import { quizShape } from 'propShapes/quiz'
import { modes } from 'components/containers/QuizEditor/reducer'
import { pcPanelModeIds } from 'components/containers/QuizEditor/constants'

import FormFields from 'components/presentational/EditorPage/FormFields'
import FormControls from 'components/presentational/EditorPage/FormControls'
import EditorControls from 'components/presentational/EditorPage/EditorControls'

import UnselectedMessage from './UnselectedMessage'

import Wrapper from './Wrapper'
import StyledForm from './StyledForm'

const Editor = ({
  mode,
  original,
  buffer,
  panelModes,
  onSelectNewItem,
  onOpenPreview,
  ...rest
}) => {
  const isNew = mode === modes.NEW
  const isPristine = equals(original, buffer)
  const hasSelection = mode !== modes.UNSELECTED
  const panelModesForType = type => (
    type === 'pc'
      ? filter(pm => contains(pm.id, pcPanelModeIds), panelModes)
      : panelModes)

  return (
    <Wrapper>
      <EditorControls
        {...{
          onSelectNewItem, onOpenPreview, isNew, hasSelection,
        }}
      />
      {mode === modes.UNSELECTED
        ? <UnselectedMessage />
        :
        <StyledForm
          {...{
            buffer,
            isNew,
            isPristine,
            ...rest,
          }}
        >
          <FormFields panelModes={panelModesForType(original.type)} />
          <FormControls />
        </StyledForm>
      }
    </Wrapper>
  )
}


Editor.propTypes = {
  mode: pt.string.isRequired,
  original: quizShape,
  buffer: quizShape,
  panelModes: pt.arrayOf(pt.shape({})).isRequired,
  onUpdateBuffer: pt.func.isRequired,
  onSelectNewItem: pt.func.isRequired,
  onOpenPreview: pt.func.isRequired,
  onCreate: pt.func.isRequired,
  onUpdate: pt.func.isRequired,
  onDelete: pt.func.isRequired,
}

Editor.defaultProps = {
  original: undefined,
  buffer: undefined,
}

export default Editor
