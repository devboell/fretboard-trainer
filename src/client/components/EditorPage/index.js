import React from 'react'

import ListEditor from 'components/ListEditor'


const EditorPage = () =>
  <div>
    Editor Page
    <ListEditor />
  </div>

EditorPage.propTypes = {}

EditorPage.defaultProps = {}

export default EditorPage
