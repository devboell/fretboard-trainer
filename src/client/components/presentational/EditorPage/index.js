import React from 'react'

import ListEditor from 'components/presentational/EditorPage/ListEditor'


const EditorPage = () =>
  <div>
    Editor Page
    <ListEditor />
  </div>

EditorPage.propTypes = {}

EditorPage.defaultProps = {}

export default EditorPage
