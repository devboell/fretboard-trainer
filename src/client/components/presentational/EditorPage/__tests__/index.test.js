import React from 'react'

import EditorPage from '../index'


it('EditorPage component, snapshot', () => {
  const wrapper = shallow(<EditorPage />)
  expect(wrapper).toMatchSnapshot()
})
