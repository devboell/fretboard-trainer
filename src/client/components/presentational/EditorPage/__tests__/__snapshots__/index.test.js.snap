// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage component, snapshot 1`] = `
<div>
  <Connect(ListEditor) />
</div>
`;
